Будем осуществлять работу с непростым [набором данных](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.csv) о состоянии здоровья лошадей, испытывающих кишечные колики. 

#### Задание 1. Базовое изучение

Изучить представленный набор данных на основе [описания его столбцов](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.names) и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

#### Задание 2. Работа с выбросами

В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

#### Задание 3. Работа с пропусками

Рассчитать количество выбросов для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

In [2]:
import pandas as pd

col_names = ['surgery', 'age', 'id', 'temp', 'pulse', 'rrate', 'pain', 'abdomen', 'outcome']
df = pd.read_csv('horse_data.csv', header=None, na_values='?', usecols=[0, 1, 2, 3, 4, 5, 10, 17, 22], names=col_names)
df.head()

surgery  age       id  temp  pulse  rrate  pain  abdomen  outcome
0      2.0    1   530101  38.5   66.0   28.0   5.0      5.0      2.0
1      1.0    1   534817  39.2   88.0   20.0   3.0      2.0      3.0
2      2.0    1   530334  38.3   40.0   24.0   3.0      1.0      1.0
3      1.0    9  5290409  39.1  164.0   84.0   2.0      NaN      2.0
4      2.0    1   530255  37.3  104.0   35.0   NaN      NaN      2.0

In [3]:
for col in df.columns:
    print(f"{col} - {df[col].mode()[0]}") 

surgery - 1.0
age - 1
id - 527544
temp - 38.0
pulse - 48.0
rrate - 20.0
pain - 3.0
abdomen - 5.0
outcome - 1.0


In [4]:
df.describe()

surgery         age            id        temp       pulse  \
count  299.000000  300.000000  3.000000e+02  240.000000  276.000000   
mean     1.397993    1.640000  1.085889e+06   38.167917   71.913043   
std      0.490305    2.173972  1.529801e+06    0.732289   28.630557   
min      1.000000    1.000000  5.184760e+05   35.400000   30.000000   
25%      1.000000    1.000000  5.289040e+05   37.800000   48.000000   
50%      1.000000    1.000000  5.303055e+05   38.200000   64.000000   
75%      2.000000    1.000000  5.347275e+05   38.500000   88.000000   
max      2.000000    9.000000  5.305629e+06   40.800000  184.000000   

            rrate       pain     abdomen     outcome  
count  242.000000  245.00000  182.000000  299.000000  
mean    30.417355    2.95102    3.692308    1.551839  
std     17.642231    1.30794    1.491725    0.737187  
min      8.000000    1.00000    1.000000    1.000000  
25%     18.500000    2.00000    2.000000    1.000000  
50%     24.500000    3.00000    4.000000    1.000000  
75%     36.000000    4.00000    5.000000    2.000000  
max     96.000000    5.00000    5.000000    3.000000

Выводы:
большинство лошадей взрослые
большинство были прооперированы
температуры в среднем 38.2 
пульс в основном 65 
по болевому синдрому распределение одинаковое
более половины выжило

In [5]:
q1 = df.temp.quantile(0.25)
q3 = df.temp.quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)

outliers = df[(df.temp > upper_bound) | (df.temp < lower_bound)].sort_values('temp')
print(outliers)

     surgery  age      id  temp  pulse  rrate  pain  abdomen  outcome
44       1.0    1  535407  35.4  140.0   24.0   4.0      5.0      3.0
141      2.0    1  522979  36.0   42.0   30.0   NaN      NaN      2.0
238      2.0    1  528702  36.1   88.0    NaN   3.0      4.0      3.0
80       1.0    1  527518  36.4   98.0   35.0   4.0      4.0      2.0
118      1.0    1  533983  36.5   78.0   30.0   5.0      NaN      1.0
298      1.0    1  530612  36.5  100.0   24.0   3.0      4.0      1.0
251      2.0    1  527940  36.6   42.0   18.0   1.0      5.0      2.0
99       2.0    1  530002  39.6  108.0   51.0   2.0      5.0      1.0
75       1.0    9  534092  39.7  100.0    NaN   2.0      NaN      3.0
20       1.0    1  530157  39.9   72.0   60.0   5.0      4.0      1.0
281      2.0    1  530624  40.0   78.0    NaN   2.0      1.0      2.0
54       2.0    1  529461  40.3  114.0   36.0   2.0      5.0      3.0
91       2.0    1  529461  40.3  114.0   36.0   2.0      5.0      2.0
259      1.0    1  5

In [6]:
q1 = df.pulse.quantile(0.25)
q3 = df.pulse.quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)

outliers = df[(df.pulse > upper_bound) | (df.pulse < lower_bound)].sort_values('pulse')
print(outliers)

     surgery  age       id  temp  pulse  rrate  pain  abdomen  outcome
41       2.0    9  5288249  39.0  150.0   72.0   NaN      NaN      1.0
275      1.0    9  5297159  38.8  150.0   50.0   5.0      NaN      2.0
55       1.0    9  5282839  38.6  160.0   20.0   3.0      NaN      2.0
3        1.0    9  5290409  39.1  164.0   84.0   2.0      NaN      2.0
255      1.0    9  5294539  38.8  184.0   84.0   4.0      NaN      2.0


Выбросы присутствуют в данных температуры, пульса, дыхания, но могут быть признаны естественными для данной выборки.
Кроме того, в графе возраст ошибка:  9 вместо 2, так же можно признать это выбросом.

In [7]:
for col in df.columns:
    pct_missing = df[col].isnull().mean()
    print(f'{col} - {pct_missing :.1%}')

surgery - 0.3%
age - 0.0%
id - 0.0%
temp - 20.0%
pulse - 8.0%
rrate - 19.3%
pain - 18.3%
abdomen - 39.3%
outcome - 0.3%


In [9]:
df['pain'].fillna('no_info', inplace=True)
df['abdomen'].fillna('no_info', inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 9 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   surgery  299 non-null    float64
 1   age      300 non-null    int64  
 2   id       300 non-null    int64  
 3   temp     240 non-null    float64
 4   pulse    276 non-null    float64
 5   rrate    242 non-null    float64
 6   pain     300 non-null    object 
 7   abdomen  300 non-null    object 
 8   outcome  299 non-null    float64
dtypes: float64(5), int64(2), object(2)
memory usage: 21.2+ KB


In [10]:
df.dropna(inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 206 entries, 0 to 299
Data columns (total 9 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   surgery  206 non-null    float64
 1   age      206 non-null    int64  
 2   id       206 non-null    int64  
 3   temp     206 non-null    float64
 4   pulse    206 non-null    float64
 5   rrate    206 non-null    float64
 6   pain     206 non-null    object 
 7   abdomen  206 non-null    object 
 8   outcome  206 non-null    float64
dtypes: float64(5), int64(2), object(2)
memory usage: 16.1+ KB


Пропуски в surgery и outcome - возможно являются ошибками ввода, можно исключить такие строки.
Числовые temp, pulse, rrate аналогично нужно удалить.